### ISSUE TITLE
[Omniscia audit] Cross-Chain Replay Attack Susceptibility

### ISSUE THREAD

**Issue by hiroshi-yamamoto-dublr** [dublr/dublr#10]
## OTI-01M: Cross-Chain Replay Attack Susceptibility

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Language Specific](/appendix/finding-types#language-specific) | <img class="o-severity o-minor" src="https://omniscia.io/report-assets/minor.png" /> | [OmniTokenInternal.sol:L55-L63](https://github.com/dublr/dublr/blob/b84ea8edd8c1ab6a3f1abbc34467fa2ce4b21526/contracts/main/OmniToken/OmniTokenInternal.sol#L55-L63) |

### Description:

In the current system, the domain separator is computed only once during the construction of the contract thus rendering it prone to cross-chain replay attacks as when an EVM blockchain forks its contract and storage space remains the same whilst the `block.chainid` member changes.

### Impact:

All signatures generated for the main chain will be replayable in a one-to-one fashion on the forked chain due to the `chainid` being identical between the two chains.

### Example:

```solidity
// Initialize EIP712 domain separator for EIP2612 permit API
DOMAIN_SEPARATOR = keccak256(
    abi.encode(
        // The value of
        // keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)")
        0x8b73c3c69bb8fe3d512ecc4cf759cc79239f7b179b0ffacaa9a75d522b39400f,
        keccak256(bytes(name)),
        keccak256(bytes(version)),
        block.chainid,
        address(this)));
```

### Recommendation:

We advise a similar approach to [OpenZeppelin's `draft-EIP712.sol` implementation](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/draft-EIP712.sol) to be utilized instead whereby the domain separator is computed during construction, cached to storage and solely utilized when the `block.chainid` member matches the cached `chainid` that was used during construction.

**Comment by hiroshi-yamamoto-dublr** [comment]
This is a problematic recommendation, because EIP2612 clearly [specifies](https://eips.ethereum.org/EIPS/eip-2612) that `DOMAIN_SEPARATOR()` should be an `external view` function of a compliant contract:

```
function DOMAIN_SEPARATOR() external view returns (bytes32)
```

This is part of the EIP2612 API. I suppose that means that the domain separator must always be dynamically generated, so that it incorporates the `chainId` of the caller?

**Comment by hiroshi-yamamoto-dublr** [comment]
Solved in bf2bba2723256a13ae8ecb45d8d1f17cd7865a6d by changing `DOMAIN_SEPARATOR` from a field into a function, and dynamically generating the separator.

**Comment by hiroshi-yamamoto-dublr** [comment]
PS there is no room left in the contract for a caching mechanism (it's at the code size limit).
